### PR TITLE
steam: fix 32bit driver check, use 32bit compatible ldd, make pressure-vessel better

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/games/steam/fhsenv.nix [hunk 1]
```diff
@@ -271,10 +271,10 @@ in buildFHSEnv rec {
   '' + extraProfile;
 
   runScript = writeShellScript "steam-wrapper.sh" ''
-    if [ -f /host/etc/NIXOS ]; then   # Check only useful on NixOS
-      ${glxinfo-i686}/bin/glxinfo >/dev/null 2>&1
+    if [ -f /etc/NIXOS ]; then   # Check only useful on NixOS
+      ${glxinfo-i686}/bin/glxinfo 2>&1 | grep -q Error
       # If there was an error running glxinfo, we know something is wrong with the configuration
-      if [ $? -ne 0 ]; then
+      if [ $? -eq 0 ]; then
```
Comment (abysssol) [inline pkgs/games/steam/fhsenv.nix:273]: > glxinfo does not write a proper error code on errors 🤦

Does this mean that glxinfo's exit code is 0 even when it encounters unrecoverable errors?

File: pkgs/games/steam/fhsenv.nix [hunk 2]
```diff
@@ -223,9 +222,7 @@ in buildFHSEnv rec {
     zlib
     atk
     cairo
-    freetype
     gdk-pixbuf
-    fontconfig
```
Comment (abysssol) [inline pkgs/games/steam/fhsenv.nix:228]: > drop game specific libraries now duplicated with steam own

To be clear, this means that steam has added these libraries to it's runtime?

**Review by abysssol — APPROVED** [review]
I'm not knowledgeable on steam's package, but everything looks good.

File: pkgs/games/steam/fhsenv.nix [hunk 3]
```diff
@@ -303,6 +299,7 @@ in buildFHSEnv rec {
   '' + args.extraPreBwrapCmds or "";
 
   extraBwrapArgs = [
+    "--bind /etc/NIXOS /etc/NIXOS" # required 32bit driver check in runScript
```
Comment (tejing1) [inline pkgs/games/steam/fhsenv.nix:302]: Wouldn't this fit better in `etcBindEntries` in the `buildFHSEnv` code?

File: pkgs/games/steam/fhsenv.nix [hunk 4]
```diff
@@ -23,9 +23,8 @@ let
     # Errors in output without those
     pciutils
     # run.sh wants ldconfig
-    glibc.bin
+    glibc_multi.bin
     # Games' dependencies
-    xorg.xrandr
```
Comment (tejing1) [inline pkgs/games/steam/fhsenv.nix:28]: What's the argument for xrandr no longer being necessary? How do we know some games don't still depend on it? Or is it included with steam's runtime as well?

File: pkgs/games/steam/fhsenv.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/games/steam/fhsenv.nix:273]: Exactly 😅

File: pkgs/games/steam/fhsenv.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/games/steam/fhsenv.nix:228]: We already have them in the list above.

File: pkgs/games/steam/fhsenv.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/games/steam/fhsenv.nix:302]: I am not sure if this is something we should universally do. Maybe?

File: pkgs/games/steam/fhsenv.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/games/steam/fhsenv.nix:28]: I checked it against the flatpak and the Debian package and they both don't require it. If a specific game requires it, we should work with overwrites instead.

**Comment by SuperSandro2000** [comment]
I want to bring the closure size reduction for steam-small back but I kinda ran into a general issue which also exists on current nixos-unstable https://github.com/ValveSoftware/steam-runtime/issues/684 and before that is not being fixed, I think this PR will stay on DRAFT to not make more damage.

File: pkgs/games/steam/fhsenv.nix (hunk 4 above)
Comment (tejing1) [reply pkgs/games/steam/fhsenv.nix:28]: If it's not in that flatpak, that is a pretty good indicator. Good enough for me.

Comment (SuperSandro2000) [reply pkgs/games/steam/fhsenv.nix:28]: 🤦🏼 derp, it is in the flatpak and they install it explcitily. Going to drop this again.

File: .git-blame-ignore-revs [hunk 5]
```diff
@@ -153,3 +153,6 @@ bdfde18037f8d9f9b641a4016c8ada4dc4cbf856
 
 # nixos/ollama: format with nixfmt-rfc-style (#329561)
 246d1ee533810ac1946d863bbd9de9b525818d56
+
+# steam: cleanup (#216972)
+ad815aebfbfe1415ff6436521d545029c803c3fb
```
Comment (K900) [inline .git-blame-ignore-revs:158]: Isn't necessary anymore, is it?

File: pkgs/build-support/build-fhsenv-bubblewrap/buildFHSEnv.nix [hunk 6]
```diff
@@ -15,6 +15,10 @@
 , extraBuildCommands ? ""
 , extraBuildCommandsMulti ? ""
 , extraOutputsToInstall ? []
+# Whether to symlink /usr/lib* or copy files directly
+# This is mainly a workaround for steam pressure-vessel bug tracked at
```
Comment (K900) [inline pkgs/build-support/build-fhsenv-bubblewrap/buildFHSEnv.nix:19]: Giga nit: Steam, capitalized. Also, might be worth reporting at https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/issues, which is a bit more targeted?

File: pkgs/games/steam/fhsenv.nix (hunk 3 above)
Comment (K900) [reply pkgs/games/steam/fhsenv.nix:307]: Maybe just kill the check instead?

File: pkgs/build-support/build-fhsenv-bubblewrap/buildFHSEnv.nix (hunk 6 above)
Comment (SuperSandro2000) [reply pkgs/build-support/build-fhsenv-bubblewrap/buildFHSEnv.nix:19]: I've read somewhere in their docs to better report issues on the GitHub. Probably that they can manage things more easily and they had similar issues in that repo in the past.

File: .git-blame-ignore-revs (hunk 5 above)
Comment (SuperSandro2000) [reply .git-blame-ignore-revs:158]: This is an old commit that annoyed me while debugging.

Comment (K900) [reply .git-blame-ignore-revs:158]: Oh, OK.

**Comment by K900** [comment]
Overall I really want to have a response from Valve before we move ahead with this hack. This would inflate the fhsenv size massively...

**Comment by SuperSandro2000** [comment]
Yeah, fine for me. I think my initial issue was caused by corrupted files for whatever reason. It also works without this patch but the runtime check run through pressure-vessel reports pots of issues which is probably not great.

**Comment by SuperSandro2000** [comment]
Yes, the steam package is now kinda fat :(

```
steam: +6069086.3 KiB
```

**Comment by SuperSandro2000** [comment]
Yep

**Comment by nyabinary** [comment]
> So are we blocked on the Valve folks shipping the pressure-vessel fix here?

It's now on client_beta (the fix) for SLR-3.0 https://github.com/ValveSoftware/steam-runtime/issues/684#issuecomment-2286115991

File: pkgs/games/steam/fhsenv.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/games/steam/fhsenv.nix:303]: It's a good hint for people to properly use the module. We can kill it later if it causes more pain.

**Pull request merged**